### PR TITLE
Ariba

### PR DESCRIPTION
**Pull request by alexwolson**
I have finished a fairly significant refactoring of the Ariba scraper. It now runs as a single Python script (per #8) and is more platform-independent than before (certainly not dependent on being run on my system!) - see #5. The code runs faster due to waiting more flexibly for things to complete, and is better at handling errors, as well as identifying when it's reached the end. It then automatically does the post-processing of moving and extracting zip files, and deduplication.

The most relevant files are `ariba_driver.py`, which overrides the driver object with some additional useful methods for this context, `filemanage.py`, which contains the methods that are run after the scraper is finished, and `rfp_scraper.py`, which handles the main scraping.

### CODE REVIEW

**Comment by Arpanio** [comment]
This is fire! I'll do the review and run through Ubuntu tomorrow.

Everyone else, please feel free to chime in as well.

File: scrapers/ariba/ariba_driver.py
Comment (sonalranjit) [inline scrapers/ariba/ariba_driver.py:63]: What is the significance of the the `key` parameter in the url, wondering if it would be a better practice to not hardcode urls inline in the code.

File: scrapers/ariba/rfp_scraper.py
Comment (sonalranjit) [inline scrapers/ariba/rfp_scraper.py:12]: Probably better if we download to a relative `downloads` directory which can be added to `.gitignore`, will be easier to have everything in one place vs. navigating to home/Downloads

**Comment by Arpanio** [comment]
Really awesome refactor! It's way easier to follow and understand now.

I had to make the following minor changes to get things working:

1. Created the files `password.key` and `username.key` using my Ariba credentials. We could keep empty files and have a manual mechanism to populate those during deployment.
2. Created a `data` folder under the `toronto-bids` project root. Without this, I'd get 
> [Errno 2] No such file or directory: '/home/arpanio/projects/toronto-bids/data/Doc3528836035.html'
3. Created a `downloads` folder under the project root as well as changed line 12 of `rfp_scraper.py` to `DOWNLOAD_DIRECTORY = Path.cwd() / 'downloads'`. Without this change, I'd get
> [Errno 2] No such file or directory: '/home/arpanio/Downloads'

At the moment, the scraper gets stuck after a while, at [this URL](https://s1.ariba.com/Sourcing/Main/aw?awh=r&awssk=vjBfWuGx&realm=toronto&passwordadapter=SourcingSupplierUser#b0) where it should download stuff. I haven't been able to debug that yet. 

Another behaviour that I observed is that the scraper opens up chrome, logs in, downloads data from an RFP, and then closes. Then the cycle starts from the beginning for the next RFP. I'm cool with this behaviour, but just want to highlight that this is how it is on Ubuntu.

Since this is a large PR, I suggest that we take some time to go through the code and create 2 tiers of issues. Stylistic changes and bugs. If there are 0 bugs, we can merge to main and work on the stylistic changes through other issues. Of course, it may be best to close both stylistic changes as well as bugs at one go, but stylistic changes can be numerous and result in a death-by-a-thousand-cuts situation.

File: scrapers/ariba/ariba_driver.py
Comment (Arpanio) [reply scrapers/ariba/ariba_driver.py:63]: That's the key to the Toronto City side of things on Ariba. I agree, we should store these elements in a config file.

File: scrapers/ariba/rfp_scraper.py
Comment (sonalranjit) [reply scrapers/ariba/rfp_scraper.py:12]: hmmm there should be an option to add browser options when creating the webdriver to specify the downloads directory. It shouldn't be on the Ariba's end. I can try that out when i get a chance to run the scripts.

Comment (alexwolson) [reply scrapers/ariba/rfp_scraper.py:12]: Seems like we can do something like

`from selenium.webdriver.chrome.options import Options

# Create an instance of Options
chrome_options = Options()

# Set the download directory
download_directory = "/path/to/your/desired/download/directory"

# Add an argument to set the default download directory
chrome_options.add_experimental_option(
    "prefs",
    {
        "download.default_directory": download_directory,
        "download.prompt_for_download": False,
        "download.directory_upgrade": True,
        "safebrowsing.enabled": True,
    },
)

# Create a WebDriver instance with the specified options
driver = webdriver.Chrome(options=chrome_options)`

Comment (sonalranjit) [reply scrapers/ariba/rfp_scraper.py]: Yup i've updated the python code to add the chrome option, i can verify it's downloading to the `REPO_DIRECOTRY/downloads` now. I've also added a line to created the `data` and `downloads` directory if it doesn't exist. please try it out and let me know

Comment (alexwolson) [reply scrapers/ariba/rfp_scraper.py]: Tested and working on Ubuntu 23.04, and MacOS 13.3. I had to make two changes - the download parameter has to be passed if we re-make the driver after an error, and I noticed the login behaviour was slightly different on MacOS so I added a couple of lines to accept that.

Comment (sonalranjit) [reply scrapers/ariba/rfp_scraper.py]: LGTM

File: scrapers/ariba/environment_ubuntu_2304.yml
Comment (sonalranjit) [inline scrapers/ariba/environment_ubuntu_2304.yml:1]: Is this necessary ? i can't imagine the env being that different from mac to ubuntu. i was able to run the `rfp_scrapyer.py` by just doing a ```source venv/bin/activate
pip install -r requirements
python scrapers/ariba/rfp_scrapers.py```

Comment (alexwolson) [reply scrapers/ariba/environment_ubuntu_2304.yml:1]: probably not. I’m curious to figure out why Arpan’s Ubuntu setup is having trouble with this code but mine isnt.

File: scrapers/ariba/ariba_driver.py
Comment (sonalranjit) [reply scrapers/ariba/ariba_driver.py]: okay i've moved this into a `rfp_scraper_config.json`

**Comment by sonalranjit** [comment]
> Pulled in the new changes and tested (Y'all are super fast!). Unfortunately, there's still an issue with Chrome suddenly stopping, crashing, and then resuming from the start.
> 
> I tried 2 approaches:
> 
> 1. Did a `git pull` in the same local copy that I was working on yesterday.
> 2. Did a fresh `git clone` into a new directory, setup a virtual environment and installed from requirements (didn't use the make setup), and ran `python3.9 scrapers/ariba/rfp_scraper.py`
> 
> Improvements from my previous comment:
> 
> 1. There's no consistent re-logging in. Once Chrome starts, login is completed and we move from 1 RFP to the next without Chrome closing. This is intended behaviour.
> 2. The `data` and `downloads` directories are created when they don't exist.
> 
> Output of run 1 (same local copy as yesterday's) is available [here](https://pastebin.com/DQjQUrW3).
> 
> * At line 82, the script stops printing while Chrome is open
> * I can freely scroll around in Chrome
> * After a while, Chrome crashes and the following stacktrace is emitted
> * A new Chrome window starts up, the script logs in, and we proceed from there
> 
> Output of run 2 (fresh `git clone`) is pretty similar. I got a [screenshot](https://imgur.com/a/zcTmN9W) as well. During this run, I decided to not change the active window and let Chrome always be visible and selected. This didn't help.
> 
> In both cases, I have `html` files in `data` but nothing in `downloads`.
> 
> It looks like an issue with some Ubuntu-specific drivers. We also want to run this in a headless setting, so it wouldn't make sense to try and debug this too hard. My suggestions are as follows:
> 
> **Issues that we should still consider in this PR:**
> 
> 1. Hardcoded URLs should be moved to a config
> 2. How is `environment_ubuntu_2304.yml` used? It has a reference to Alex's folder (`/home/alex/anaconda3/envs/bids`)
> 
> **Issues that we should pick up outside of this PR (not an exhaustive list):**
> 
> 1. Better logging - Use a logging library instead of printing to console so we can investigate issues after going live
> 2. Ensure scraping via headless Chrome works on any 1 Linux OS
> 3. It looks like we process the `main_loop` and then move files from `download` to `data`. Revisit this as there's a potential to get stuck in an infinite loop and get 0 data uploaded, even if we successfully process 99/100 RFPs and fail on the very last RFP.

okay i've removed the ubuntu environment yml and move the hardcoded URL to a base url variable because i think that won't change anytime soon.
Created a `rfp_scraper_config.json` that just holds the profile key parameter for now, but we can extend that json to hold other future configuration parameters. 

And yes i agree we should setup a proper logging, run selenium in headless because we will have to eventually.
Also as far as linting goes i'd suggest we just go with `black` let it run on PRs or get in the habit of running the lint before we commit.

**Comment by Arpanio** [comment]
Not all branches were using the key parameter. Updated that and tested.

Surprisingly, I still get `TypeError: home() missing 1 required positional argument: 'profile_key'` once we hit the except-else clause and re-initiate `Ariba`. We can pick that up in another issue though, since hitting that exception itself is a problem.

I'll let @alexwolson have a look before merging.

**Review by Arpanio — APPROVED** [review]

**Review by sonalranjit — APPROVED** [review]

**Comment by alexwolson** [comment]
Thanks so much for all your help guys. I feel like I learned a ton from this exercise. Let’s make issues for some of the points noted above, so they’re tracked for later fixes.

**Pull request merged**
